### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by rogeralsing** [antmicro/Migrant#92]
Hi,
I have  been comparing the perf of Migrant with Json.NET today and the results were a bit on the weak side.
It turned out that Migrant is about 24 times slower than Json.NET.
I do get that there is more data to handle with Migrant as it includes a lot more in the payload than json.net does.

Here is a is a screenshot of a profiling session:

![migrantperf](https://cloud.githubusercontent.com/assets/647031/9280212/15c58a4a-42bd-11e5-8a13-b42e5a650430.png)

And here is the (extremely naive) perf test I used https://gist.github.com/rogeralsing/7b57167935ce048859e7
The test is designed to simulate what we do in Akka.NET, thus the byte array stuff in there.

According to the profiling session, it does look like there is a specific method that is responsible for most of the cpu consumption, `TouchAndWriteAssemblyId` 

Are there any settings that can omit some of the expensive information from the payload or are the default settings the fastest?

**Comment by rogeralsing** [comment]
In ObjectWriter.cs - TouchAndWriteAssemblyId

```csharp
if(assemblyIndices.ContainsKey(assembly))
            {
                assemblyId = assemblyIndices[assembly];
```

Changing that into a `assemblyIndices.TryGetValue` would make it twice as fast as there is just one hash lookup instead of two (first checking existence and then getting the value)

**Comment by rogeralsing** [comment]
I've got the total time down to 0.898 sec now. it turns out that the Equality comparison for AssemblyDescriptor uses string join and some linq to generate the FullName of the assembly, the full name is then used in the equality comparison.
So just by caching the FullName of the assembly descriptor made things a ton better.

**Comment by konrad-kruczynski** [comment]
Hi roger,
first of all the way Migrant is used in the test is not the one Migrant is meant to be used, since with each new serialize it has to create all type stamps and headers. Please compare serializing 100000 objects in a loop to 100000 objects contained in an array and serialized using one Serialize call. Then, depending on the scenario you have, you either want to use open stream serialization or some collection.

**Comment by konrad-kruczynski** [comment]
Normally the heavy process you're mainly profiling now is done only once per type (so in case of 100000 instances the time it takes in negligible).

**Comment by konrad-kruczynski** [comment]
Nonetheless your findings are still interesting and will be analyzed (you can even post a pull request if you'd like to).

**Comment by konrad-kruczynski** [comment]
I've modified your test snippet according to what I've written. Look at it here: https://gist.github.com/konrad-kruczynski/2636b9c2eaa37961c38b

Please run it and tell me whether it suits your needs.

**Comment by konrad-kruczynski** [comment]
I've used your `TryGetValue` tip in ObjectIdentifier.

**Comment by konrad-kruczynski** [comment]
By temporary removing stamping (just as an experiment now, not the final version) I was able to get down from 4.1s to 0.45s on my PC (your unmodified tests). So it will definitely be a way to go.

**Comment by mateusz-holenko** [comment]
I created a separate issue on creating an option to disable type stamping: #98.

**Comment by konrad-kruczynski** [comment]
I am closing this ticket for now since its problems are analyzed in #93 and #98.
